core/server/worker/src/main/java/alluxio/worker/block/UfsInputStreamManager.java

```diff
@@ -0,0 +1,341 @@
+   * @return the acquired input stream
+   * @throws IOException if the input stream fails to open
+   */
+  public InputStream acquire(UnderFileSystem ufs, String path, long fileId, OpenOptions openOptions,
+      boolean reuse) throws IOException {
+    if (!ufs.isSeekable() || !isCachingEnabled()) {
+      // not able to cache, always return a new input stream
+      return ufs.open(path, openOptions);
+    }
+
+    // explicit cache cleanup
+    mUnderFileInputStreamCache.cleanUp();
+
+    UfsInputStreamIds resources;
+    synchronized (mFileIdToInputStreamIds) {
+      if (mFileIdToInputStreamIds.containsKey(fileId)) {
+        resources = mFileIdToInputStreamIds.get(fileId);
+      } else {
+        resources = new UfsInputStreamIds();
+        mFileIdToInputStreamIds.put(fileId, resources);
+      }
+    }
+
+    synchronized (resources) {
+      long nextId = UNAVAILABLE_RESOURCE_ID;
+      CachedSeekableInputStream inputStream = null;
+      if (reuse) {
+        // find the next available input stream from the cache
+        for (long id : resources.availableIds()) {
+          inputStream = mUnderFileInputStreamCache.getIfPresent(id);
+          if (inputStream != null) {
+            nextId = id;
+            LOG.debug("Reused the under file input stream resource of {}", nextId);
+            // for the cached ufs instream, seek to the requested position
+            inputStream.seek(openOptions.getOffset());
+            break;
+          }
+        }
+      }
+      // no cached input stream is available, open a new one
+      if (nextId == UNAVAILABLE_RESOURCE_ID) {
+        nextId = IdUtils.getRandomNonNegativeLong();
+        final long newId = nextId;
+        try {
+          inputStream = mUnderFileInputStreamCache.get(nextId, () -> {
+            SeekableUnderFileInputStream ufsStream = (SeekableUnderFileInputStream) ufs.open(path,
+                OpenOptions.defaults().setOffset(openOptions.getOffset()));
+            LOG.debug("Created the under file input stream resource of {}", newId);
+            return new CachedSeekableInputStream(ufsStream, newId, fileId, path);
+          });
+        } catch (ExecutionException e) {
+          LOG.warn("Failed to retrieve the cached UFS instream");
+          // fall back to a ufs creation.
+          return ufs.open(path, OpenOptions.defaults().setOffset(openOptions.getOffset()));
+        }
+      }
+
+      // mark the input stream id as acquired
+      resources.acquire(nextId);
+      return inputStream;
+    }
+  }
+
+  private boolean isCachingEnabled() {
+    return Configuration.getBoolean(PropertyKey.WORKER_UFS_INSTREAM_CACHE_ENABLE);
+  }
+
+  /**
+   * The metadata of the input streams associated with an under storage file that tracks which input
+   * streams are in-use or available. Each input stream is identified by a unique id.
+   */
+  @ThreadSafe
+  private static class UfsInputStreamIds {
+    private final Set<Long> mInUse;
```

Comment:
Why do we need to keep track of in-use ids?

Reply:
I think it's useful to track the in-use resources in a central place, easier for auditing and state checking.

Reply:
Having more information available is always better, but it looks like `mInUse` costs us a significant amount of complexity. If we just keep track of `mAvailable`, could we just use a `ConcurrentHashSet`  instead of this class?

Reply:
Discussed offline. Removing `mInUse` won't improve the concurrency design, as the same locking is still necessary on the Hashset.